src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs

```diff
@@ -3938,36 +3955,82 @@ private static void AnalyzeReturnType(IMethodSymbol oldMethod, IMethodSymbol new
+                    {
+                        hasReturnTypeChange = true;
+                    }
+                    else
+                    {
+                        rudeEdit = RudeEditKind.ChangingTypeNotSupportedByRuntime;
+                    }
+                }
                 else
                 {
                     rudeEdit = RudeEditKind.TypeUpdate;
                 }
             }
         }
 
-        private static void AnalyzeReturnType(IEventSymbol oldEvent, IEventSymbol newEvent, ref RudeEditKind rudeEdit, ref bool hasGeneratedReturnTypeAttributeChange)
+        private void AnalyzeReturnType(IEventSymbol oldEvent, IEventSymbol newEvent, EditAndContinueCapabilitiesGrantor capabilities, ref RudeEditKind rudeEdit, ref bool hasGeneratedReturnTypeAttributeChange, ref bool hasReturnTypeChange, CancellationToken cancellationToken)
         {
             if (!ReturnTypesEquivalent(oldEvent, newEvent, exact: true))
             {
                 if (ReturnTypesEquivalent(oldEvent, newEvent, exact: false))
                 {
                     hasGeneratedReturnTypeAttributeChange = true;
                 }
+                else if (oldEvent.ContainingType.IsDelegateType())
+                {
+                    rudeEdit = RudeEditKind.TypeUpdate;
+                }
+                else if (AllowsDeletion(newEvent))
+                {
+                    if (CanAddNewMember(newEvent, capabilities, cancellationToken))
+                    {
+                        hasReturnTypeChange = true;
+                    }
+                    else
+                    {
+                        rudeEdit = RudeEditKind.ChangingTypeNotSupportedByRuntime;
+                    }
+                }
                 else
                 {
                     rudeEdit = RudeEditKind.TypeUpdate;
                 }
             }
         }
 
-        private static void AnalyzeReturnType(IPropertySymbol oldProperty, IPropertySymbol newProperty, ref RudeEditKind rudeEdit, ref bool hasGeneratedReturnTypeAttributeChange)
+        private void AnalyzeReturnType(IPropertySymbol oldProperty, IPropertySymbol newProperty, EditAndContinueCapabilitiesGrantor capabilities, ref RudeEditKind rudeEdit, ref bool hasGeneratedReturnTypeAttributeChange, ref bool hasReturnTypeChange, CancellationToken cancellationToken)
         {
             if (!ReturnTypesEquivalent(oldProperty, newProperty, exact: true))
             {
                 if (ReturnTypesEquivalent(oldProperty, newProperty, exact: false))
                 {
                     hasGeneratedReturnTypeAttributeChange = true;
                 }
+                else if (oldProperty.ContainingType.IsDelegateType())
+                {
+                    rudeEdit = RudeEditKind.TypeUpdate;
+                }
+                else if (ReturnTypeRefsAndModifiersEquivalent(oldProperty, newProperty, exact: true) &&
```

Comment:
>ReturnTypeRefsAndModifiersEquivalent [](http://example.com/codeflow?start=25&length=36)

Why check refs and modifiers here and not with method return type?

Reply:
Do we have tests that cover changing ref/modifiers?

Reply:
This is something I still need to check. The test that hits this is actually invalid (its a ref property in a class, which doesn't compile), so out of caution I left the behavior as is, because with properties there is a field which changes type, and I don't understand the implications of that enough.

Reply:
> Do we have tests that cover changing ref/modifiers?

Yes, for both properties and methods. If there was a test for it, there is no way I would have thought to check for this :)

Reply:
Okay, turns out ref properties can't be auto-properties, so the field worry is not there, so there is probably no point blocking this.

Also while adding tests for this in EditAndContinueTests I have discovered I have more work to do on this, as we currently crash during emit if properties or indexers are updated. My bad.

Reply:
Opened https://github.com/dotnet/roslyn/pull/63642 to revert parameter type changes on properties. Will bring them back in this PR, once I've worked out the changes needed in DeltaMetadataWriter etc.

Reply:
Okay, worked out the issue. Seems I took the "properties are just two methods in a trenchcoat" too far, and whilst that works for deletes, because we only care about updating accessor to throw, but other changes we need to issue `Insert` edits for the entire property, so that we can correct emit all of the associated gubbins.